File: python/leetcode074.py
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return False
        ls = len(matrix)
        if ls == 1:
            return self.binary(matrix[0], target)
        i = 0
        if matrix[i][0] > target:
            return False
        while i < ls:
            if matrix[i][0] > target:
                break
            i += 1
        row = i-1
        return self.binary(matrix[row], target)


    def binary(self,l, target):
        ls= len(l)
        start = 0
        end = ls-1
        while end >= start:
            mid = (start+end)//2
            if l[mid] == target:
                return True
            elif l[mid] < target:
                start = mid+1
            else:
                end = mid - 1
        return False


if __name__ == "__main__":
    s = Solution()
    print(s.searchMatrix([[1],[3]],1))

File: leetcodec++/leetcode074.cpp
//
// Created by 黄敏辉 on 2019-07-22.
//


class Solution {
public:
    bool searchMatrix(vector<vector<int> >& matrix, int target)
    {
        int m = matrix.size();
        if (!m) return false;
        int n = matrix[0].size();
        if (!n) return false;

        int l=-1, r=m*n-1;
        while(r-l>1)
        {
            int mid = l + (r-l)/2;
            if (matrix[mid/n][mid%n]==target) return true;
            else if (matrix[mid/n][mid%n]>target) r=mid;
            else l=mid;
        }
        if (matrix[r/n][r%n]==target) return true;
        else return false;
    }
};